src/ArrayComponents.ml

```diff
@@ -0,0 +1,53 @@
+open Base
+
+open Expr
+
+let all = [
+  {
+    name = "select";
+    codomain = Type.TVAR ("'b");
+    domain = [Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b"); Type.TVAR "'a"];
+    is_argument_valid = (function
+                         | _ -> true);
+    evaluate = (function [@warning "-8"]
+                [Value.Array (_, _, elems, default_val) ; key]
+                -> match List.Assoc.find elems ~equal:Value.equal key with
+                   | None -> default_val
+                   | Some value -> value);
+    to_string = (fun [@warning "-8"] [a ; b] -> "(select " ^ a ^ " " ^ b ^ ")");
+    global_constraints = (fun _ -> [])
+  }
+  ;
+  {
+    name = "store";
+    codomain = Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b");
+    domain = [Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b"); Type.TVAR "'a"; Type.TVAR "'b"];
+    is_argument_valid = (function
+                         | _ -> true);
+    evaluate = (function [@warning "-8"]
+                [Value.Array (key_type, val_type, elems, default_val) ; key; value]
+                -> Value.Array (key_type,val_type, (key, value)::elems,default_val));
+    to_string = (fun [@warning "-8"] [a ; b ; c] -> "(store " ^ a ^ " " ^ b ^ " " ^ c ^ ")");
+    global_constraints = (fun _ -> [])
+  }
+
+  ;
+  {
+    name = "equal";
+    codomain = Type.BOOL;
+    domain = [Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b"); Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b")];
+    is_argument_valid = (function
+                         | _ -> true);
+    evaluate = (function [@warning "-8"]
+                [Value.Array (a_key_type, a_val_type, a_elems, a_default_val) ; Value.Array (b_key_type, b_val_type, b_elems, b_default_val)]
+                -> let dedup_a = (Utils.dedup_and_sort_keep_first a_elems ~compare:(fun (k1,_) (k2,_) -> Poly.compare k1 k2))
+                   in  let dedup_b = (Utils.dedup_and_sort_keep_first b_elems ~compare:(fun (k1,_) (k2,_) -> Poly.compare k1 k2))
```

Comment:
I realized why JaneStreet doesn't provide a `which_to_keep` option for the `dedup_and_sort` function.

The `sort` function provides no guarantee on the order of _equal_ elements -- in our case when k1 == k2. When the order can be arbitrary, `First` and `Last` don't mean anything, hence they don't even provide the `which_to_keep` option. In our case, however, we need a stronger guarantee -- we want the order of _equal_ elements to be preserved, so that we can ignore everything but the last store (i.e. the `First` element). The `stable_sort` function provides this stronger guarantee.

What we really need is `dedup_and_stable_sort`. Concretely, the above two calls should be:
```ocaml
let dedup_a = List.dedup_and_stable_sort a_elems ~which_to_keep:`First ~compare:array_key_compare
in  let dedup_b = List.dedup_and_stable_sort b_elems ~which_to_keep:`First ~compare:array_key_compare
```

And you can define:
```ocaml
let array_key_compare (k1,_) (k2,_) = Value.compare k1 k2
```
at the top, before `let all = ...`

Also add `open Utils` at the top (after `Expr`) and see my new comment in `Utils.ml`